In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 500)

In [2]:
reviews=pd.read_csv('reviews_details.csv')
listing=pd.read_csv('listings_details.csv')

In [3]:
listing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20030 entries, 0 to 20029
Data columns (total 96 columns):
id                                  20030 non-null int64
listing_url                         20030 non-null object
scrape_id                           20030 non-null int64
last_scraped                        20030 non-null object
name                                19992 non-null object
summary                             19510 non-null object
space                               14579 non-null object
description                         19906 non-null object
experiences_offered                 20030 non-null object
neighborhood_overview               13257 non-null object
notes                               9031 non-null object
transit                             13635 non-null object
access                              12227 non-null object
interaction                         11974 non-null object
house_rules                         12571 non-null object
thumbnail_url           

In [4]:
listdata = []
listfeat=list(listing.columns)
for item in listfeat:
    list1=[]
    list1.append(item)
    list1.append(str(listing[item].dtype))
    list1.append(listing[item].isna().sum())
    list1.append(round((listing[item].isna().sum()/len(listing))*100,2))
    list1.append(listing[item].nunique())
    list1.append(list(listing[item].value_counts().head(2).index))
    listdata.append(list1)

desc=pd.DataFrame(listdata,columns='dataFeatures dataType null nullPct unique uniqueSample'.split(' '))
desc

dataFeatures dataType   null  nullPct  unique  \
0                                 id    int64      0     0.00   20030   
1                        listing_url   object      0     0.00   20030   
2                          scrape_id    int64      0     0.00       1   
3                       last_scraped   object      0     0.00       2   
4                               name   object     38     0.19   19503   
5                            summary   object    520     2.60   19003   
6                              space   object   5451    27.21   14227   
7                        description   object    124     0.62   19648   
8                experiences_offered   object      0     0.00       1   
9              neighborhood_overview   object   6773    33.81   12029   
10                             notes   object  10999    54.91    7270   
11                           transit   object   6395    31.93   12944   
12                            access   object   7803    38.96    9615   
13                       interaction   object   8056    40.22    9634   
14                       house_rules   object   7459    37.24   10681   
15                     thumbnail_url  float64  20030   100.00       0   
16                        medium_url  float64  20030   100.00       0   
17                       picture_url   object      0     0.00   19937   
18                    xl_picture_url  float64  20030   100.00       0   
19                           host_id    int64      0     0.00   17264   
20                          host_url   object      0     0.00   17264   
21                         host_name   object      4     0.02    5915   
22                        host_since   object      4     0.02    2829   
23                     host_location   object     39     0.19     550   
24                        host_about   object   8227    41.07    9467   
25                host_response_time   object   9483    47.34       4   
26                host_response_rate   object   9483    47.34      58   
27              host_acceptance_rate  float64  20030   100.00       0   
28                 host_is_superhost   object      4     0.02       2   
29                host_thumbnail_url   object      4     0.02   17240   
30                  host_picture_url   object      4     0.02   17240   
31                host_neighbourhood   object   5808    29.00      87   
32               host_listings_count  float64      4     0.02      43   
33         host_total_listings_count  float64      4     0.02      43   
34                host_verifications   object      0     0.00     354   
35              host_has_profile_pic   object      4     0.02       2   
36            host_identity_verified   object      4     0.02       2   
37                            street   object      0     0.00     103   
38                     neighbourhood   object   1653     8.25      44   
39            neighbourhood_cleansed   object      0     0.00      22   
40      neighbourhood_group_cleansed  float64  20030   100.00       0   
41                              city   object      4     0.02      30   
42                             state   object    127     0.63      55   
43                           zipcode   object    866     4.32    6870   
44                            market   object     42     0.21       4   
45                    smart_location   object      0     0.00      31   
46                      country_code   object      0     0.00       1   
47                           country   object      0     0.00       1   
48                          latitude  float64      0     0.00   20030   
49                         longitude  float64      0     0.00   20030   
50                 is_location_exact   object      0     0.00       2   
51                     property_type   object      0     0.00      31   
52                         room_type   object      0     0.00       3   
53                      accommodates    int64      0     0.00      15   
54       

In [5]:
listdata = []
listfeat=list(reviews.columns)
for item in listfeat:
    list1=[]
    list1.append(item)
    list1.append(str(reviews[item].dtype))
    list1.append(reviews[item].isna().sum())
    list1.append(round((reviews[item].isna().sum()/len(reviews))*100,2))
    list1.append(reviews[item].nunique())
    list1.append(list(reviews[item].value_counts().head(2).index))
    listdata.append(list1)

df_desc=pd.DataFrame(listdata,columns='dataFeatures dataType null nullPct unique uniqueSample'.split(' '))
df_desc

dataFeatures dataType  null  nullPct  unique  \
0     listing_id    int64     0     0.00   17624   
1             id    int64     0     0.00  431830   
2           date   object     0     0.00    2874   
3    reviewer_id    int64     0     0.00  411813   
4  reviewer_name   object     0     0.00   54244   
5       comments   object   530     0.12  419933   

                                        uniqueSample  
0                                   [82482, 1469096]  
1                             [111155197, 106821906]  
2                           [2017-05-28, 2018-04-02]  
3                               [12574897, 44092711]  
4                                   [David, Michael]  
5  [The host canceled this reservation 2 days bef...

In [6]:
listing=listing[listing.number_of_reviews>0]

In [7]:
listing.iloc[10]

id                                                                              42970
listing_url                                        https://www.airbnb.com/rooms/42970
scrape_id                                                              20181206172549
last_scraped                                                               2018-12-06
name                                      Comfortable room@PERFECT location + 2 bikes
summary                             A home away from home Great location Including...
space                               What I offer: great location, comfortable bed,...
description                         A home away from home Great location Including...
experiences_offered                                                              none
neighborhood_overview               My home is in the best neighborhood. Everythin...
notes                                                                             NaN
transit                             It is pretty easy 

In [8]:
new_listing=listing[['id','name','listing_url','summary','space','description','host_id','host_name','picture_url','neighbourhood_cleansed','latitude','longitude','property_type','room_type','accommodates','amenities','price','number_of_reviews','review_scores_rating','review_scores_cleanliness','review_scores_location','review_scores_checkin','review_scores_accuracy','review_scores_communication','review_scores_value']]

In [9]:
def removedollar(price):
    if ',' in str(price) :
        str_price = str(price).replace(",", "")
        return str_price.split('.')[0][1:]
    else :
        return str(price).split('.')[0][1:]
new_listing['price'] = pd.to_numeric(new_listing['price'].apply(removedollar))


In [10]:
new_listing.dropna(inplace=True)
new_listing['listing_id']=new_listing['id']
new_listing.drop(['id'], axis=1,inplace=True)
new_listing.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 12750 entries, 0 to 20007
Data columns (total 25 columns):
name                           12750 non-null object
listing_url                    12750 non-null object
summary                        12750 non-null object
space                          12750 non-null object
description                    12750 non-null object
host_id                        12750 non-null int64
host_name                      12750 non-null object
picture_url                    12750 non-null object
neighbourhood_cleansed         12750 non-null object
latitude                       12750 non-null float64
longitude                      12750 non-null float64
property_type                  12750 non-null object
room_type                      12750 non-null object
accommodates                   12750 non-null int64
amenities                      12750 non-null object
price                          12750 non-null int64
number_of_reviews              12750 non-nul

In [11]:
new_listing=new_listing.reset_index(drop=True)

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import re
def clean_text(x):
    letter = re.sub("[^a-zA-Z1-9]","_",x)
    return ' '.join(letter.split()).lower()

def new_amenties(x):
    fasilitas = str(x)[1:-1].split(',')
    fas=[]
    for item in fasilitas:
        name = clean_text(item)
        if 'missing' not in name :
            fas.append(name)
    return ' '.join(fas)
new_listing['amenities']=new_listing['amenities'].apply(new_amenties)


In [13]:
cv_res = CountVectorizer()
cv_res_hotel = cv_res.fit_transform(new_listing['amenities'])
cos_sim = cosine_similarity(cv_res_hotel)

In [14]:
df_sim=pd.DataFrame(cos_sim,index=new_listing.index,columns=new_listing.listing_id)
df_sim['loc'] = new_listing['neighbourhood_cleansed']


In [15]:
def removelistingid(x):
    if x in df_sim.columns:
        return x
    else :
        return np.nan
reviews['listing_id']=reviews['listing_id'].apply(removelistingid)
reviews.dropna(inplace=True)
reviews=reviews.sort_values(by='date',ascending=False)
list_id=[]
for item in list(reviews.listing_id.value_counts().head(300).index):
    if item in df_sim.columns:
        list_id.append(item)
len(list_id)
df_sim1=df_sim[list_id]
df_sim1['loc'] = new_listing['neighbourhood_cleansed']
df_sim1['loc']
df_sim1.shape

(12750, 301)

In [16]:
def get_recom_hotel(id_rev,loc):
#     reviews = pd.read_csv('clean_reviews.csv')
#     df_sim1 = pd.read_csv('cleansim.csv')
#     new_listing = pd.read_csv('clean_listing.csv')
#     print(df_sim1.shape)
#     print(new_listing.shape)
    index_similar=df_sim1[df_sim1['loc']==loc][(int(reviews[reviews.reviewer_id==id_rev].listing_id.iloc[0]))].sort_values(ascending=False).head(11).index
#     print(index_similar)
    return new_listing.loc[index_similar][1:]
get_recom_hotel(10952,'De Baarsjes - Oud-West')

name  \
12546              CENTRE Apartment NEAR MUSEUM DISTRICT   
10815   Typical A’dam-style apartment with lovely garden   
425                    The Best 2-Floor CANAL Apartment!   
2171                 Romantic app. Vondelpark with bikes   
2403   Quaint Outhouse in Amsterdam West  ( pvt bathr...   
7389       Beautiful City Centre Garden independent Room   
8817               Hidden gem with garden in city centre   
4701                  Unique centrally located apartment   
3549    3 room apart. with lovely garden near Vondelpark   
9447   CENTRAL Design private Apartment, walk everywh...   

                                 listing_url  \
12546  https://www.airbnb.com/rooms/29046160   
10815  https://www.airbnb.com/rooms/23444097   
425      https://www.airbnb.com/rooms/817058   
2171    https://www.airbnb.com/rooms/4339481   
2403    https://www.airbnb.com/rooms/4887616   
7389   https://www.airbnb.com/rooms/16142970   
8817   https://www.airbnb.com/rooms/19145903   
4701   https://www.airbnb.com/rooms/10095391   
3549    https://www.airbnb.com/rooms/7379564   
9447   https://www.airbnb.com/rooms/20225909   

                                                 summary  \
12546  In the city centre, near Leidseplein and very ...   
10815  Spacious apartment in the upcoming and lively ...   
425    Have Amsterdam right at your feet and watch th...   
2171   This is a sweet, romantic apartment on the sec...   
2403   This ground floor apartment in Amsterdam West ...   
7389   You will be hosted in our beautiful independen...   
8817   Great apartment with lush garden. Very suitabl...   
4701   With only 2 minutes walking distance from the ...   
3549   The apartment with private entrance, located i...   
9447   This ground floor gorgeous apartment with ensu...   

                                                   space  \
12546  The kitchenette  is fully furnished with micro...   
10815  Modern, spacious apartment. Big living room an...   
425    This apartment is brand new and spacious (130 ...   
2171   If the sun shines there's always a side of the...   
2403   The quaint bedroom comes with a four-poster be...   
7389   You will share the entrance of the house and t...   
8817   The master bedroom boasts a large king sized b...   
4701   Being a former shop that was around for more t...   
3549   Welcome to my apartment, feel at home and enjo...   
9447   This Wonderfully Decorated Deluxe Apartment wi...   

                                             description   host_id  \
12546  In the city centre, near Leidseplein and very ...    389623   
10815  Spacious apartment in the upcoming and lively ...  10474724   
425    Have Amsterdam right at your feet and watch th...   4291496   
2171   This is a sweet, romantic apartment on the sec...   3157950   
2403   This ground floor apartment in Amsterdam West ...  25142874   
7389   You will be hosted in our beautiful independen...   5785909   
8817   Great apartment with lush garden. Very suitabl...   6810286   
4701   With only 2 minutes walking distance from the ...  20804343   
3549   The apartment with private entrance, located i...    758308   
9447   This ground floor gorgeous apartment with ensu...  71257610   

              host_name                                        picture_url  \
12546               Ric  https://a0.muscache.com/im/pictures/6005b520-d...   
10815              Anne  https://a0.muscache.com/im/pictures/c69ba5d0-9...   
425       Niels & Sofia  https://a0.muscache.com/im/pictures/e182272a-5...   
2171             Annoek  https://a0.muscache.com/im/pictures/54776201/d...   
2403              Varun  https://a0.muscache.com/im/pictures/4b545bf8-8...   
7389   Alvaro And Adrie  https://a0.muscache.com/im/pictures/7d296ae4-e...   
8817                Guy  https://a0.muscache.com/im/pictures/6eaae5be-a...   
4701                Bas  https://a0.muscache.com/im/pictures/6459c6b1-c...   
3549              Kutay  https://a0.muscache.com/im/pictures/b68706

In [17]:
reviews[reviews['reviewer_name']=='Lam'].sort_values('date')
reviews[reviews['reviewer_name']=='Lam'].sort_values(by='date',ascending=False)

def removelistingid1(x):
    if x in df_sim1.columns:
        return x
    else :
        return np.nan
print(reviews.shape)
reviews['listing_id']=reviews['listing_id'].apply(removelistingid1)
reviews.dropna(inplace=True)
reviews=reviews.sort_values(by='date',ascending=False)
reviews.reviewer_id.value_counts()
print(reviews.shape)
reviews.reviewer_id.value_counts().head(10)

(354790, 6)
(82502, 6)


18024656    6
52093112    6
31635589    4
1846693     4
3281343     4
14619699    4
825755      4
5942510     4
8526157     4
56232374    4
Name: reviewer_id, dtype: int64

In [18]:
reviews.head(10)

listing_id         id        date  reviewer_id reviewer_name  \
114561   2763386.0  355928444  2018-12-06     27370265         Dante   
35089     654931.0  355925574  2018-12-06    130719768          Josh   
235846   9146453.0  355891140  2018-12-06     22463674         Peter   
162252   4934237.0  355915817  2018-12-06     32507569      Cranston   
135907   3763895.0  355717840  2018-12-05    191696819         Laila   
1726       29051.0  355704327  2018-12-05    117261720         Jonas   
37354     690757.0  355713991  2018-12-05    166848795        Leslie   
259503  11178401.0  355512015  2018-12-04     79566639          Anna   
206085   7276869.0  355484481  2018-12-04     32575647         Allen   
203910   7190115.0  355475562  2018-12-04    130417375        Wilson   

                                                 comments  
114561  Rick is a fantastic host! His apartment room i...  
35089   Had a great two night stay wish we could have ...  
235846  Akko’s place is the best spot possible in Amst...  
162252  Great space. Easy communication. Would definit...  
135907  The host was very nice and kind and she will h...  
1726    Edwin is a really cool host with a really well...  
37354   Buena ubicación pero al cuarto le hace falta i...  
259503  Said was communicative, set clear expectations...  
206085  The place was in a great location walking dist...  
203910  Wonderful service great location and very frie...

df_sim1[reviews[reviews.reviewer_id==10952].listing_id[0]]
listing[listing.id==82482]
def recom(REV_ID):
    reviews = pd.read_csv('clean_reviews.csv')
    data_login=reviews[reviews['reviewer_id']==REV_ID]
    return data_login
recom(10952).reviewer_name.iloc[0]
reviews[reviews.reviewer_id==12574897]
reviews[reviews.listing_id==82482]

new_listing.to_csv('clean_listing.csv',index=False)
reviews.to_csv('clean_reviews.csv',index=False)
df_sim1.to_csv('cleansim.csv',index=False)